### PR TITLE
stop writing header info to buffer twice

### PR DESCRIPTION
**Pull request by GodloveD**
**Description of the Pull Request (PR):**

Fixes a little bug in the way that the header is parsed in `scanDefinitionFile()`.

**This fixes or addresses the following GitHub issues:**

- fixes: #1588 

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [ ] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

**Comment by tri-adam** [comment]
@GodloveD, would you be able to write a unit test that fails before/passes after this change? I see we do have code coverage of the line you are changing, but we must not be doing enough to validate results at some level.

**Comment by GodloveD** [comment]
That's a good idea.  I need to learn how to write tests.

**Comment by bauerm97** [comment]
Hmm I don't really understand how this change can work. If we don't write into `retbuf` then nothing will be in `retbuf`, right?

**Comment by bauerm97** [comment]
@GodloveD Are you suggesting that line 72-74 take care of this and we've been returning 2x of the header this entire time?

```
 			// This line is not a section identifier
 			retbuf.Write(line)
 			retbuf.WriteString("\n")
```

**Comment by GodloveD** [comment]
@bauerm97 exactly.  Did you read #1588?

**Comment by bauerm97** [comment]
@GodloveD No, was looking at this on my phone the first time and didn't notice that there was an issue describing it :) Sorry for the confusion!

**Comment by GodloveD** [comment]
@tri-adam OK, just had a closer look at what we are and aren't testing.  So, plz correct me if I'm wrong but here is what I'm seeing.  

The error is in `scanDefinitionFile()`, but we are actually not testing at that level.  We are testing `ParseDefinitionFile()` which calls `scanDefinitionFile()` and runs the output through `doHeader()`.  The `doHeader()` function basically hides the error from `scanDefinitionFile()`, because it takes the output and enters it as keys and values into the `d.Header[]` map.  When `doHeader()` encounters duplicates, it just enters them 2x and overwrites the initial values in the map.  This covers the duplicate errors from `scanDefinitionFile()`.

Because we are testing at too high a level to catch this, I can't just add to the existing tests.  So do you want me to write a new set of tests for `scanDefinitionFile()` itself or should we just accept the higher level of testing and move on?

**Comment by bauerm97** [comment]
@GodloveD I think that writing a new test that evaluates `scanDefinitionFile()`'s behavior would be prudent

**Comment by GodloveD** [comment]
@tri-adam and @bauerm97 check this out.  Right now the test only looks at the header, but hopefully it is written in such a way that it can be expanded to the other sections as well.

File: src/pkg/build/definition_parser_deffile_test.go [hunk 1]
```diff
@@ -6,12 +6,76 @@
...
+			}
+			defer r.Close()
+
+			s := bufio.NewScanner(r)
+			s.Split(scanDefinitionFile)
+			for s.Scan() && s.Text() == "" && s.Err() == nil {
+			}
+			// debug
+			// fmt.Printf("%v\n", s.Bytes())
+			// fmt.Printf("%v\n", s.Text())
```
Comment (tri-adam) [inline src/pkg/build/definition_parser_deffile_test.go]: Remove these commented out debug lines.

File: src/pkg/build/definition_parser_deffile_test.go [hunk 2]
```diff
@@ -6,12 +6,76 @@
...
+				Header string
+			}
+			var d []DefFileSections
+			if err := json.Unmarshal(b, &d); err != nil {
+				t.Fatal("failed to unmarshal JSON:", err)
+			}
+			// debug
+			// nums := []byte(d[0].Header)
+			// fmt.Printf("%v\n", nums)
+			// fmt.Printf("%s\n", d[0].Header)
```
Comment (tri-adam) [inline src/pkg/build/definition_parser_deffile_test.go]: Same here.

**Review by tri-adam — CHANGES_REQUESTED** [review]
Sorry for the delay getting to this @GodloveD, it looks great. Thanks for adding those tests!  Just two tiny requests...

File: src/pkg/build/definition_parser_deffile_test.go (hunk 2 above)
Comment (GodloveD) [reply src/pkg/build/definition_parser_deffile_test.go]: Oh bummer.  I was afraid you were going to ask that.  I'll forget how I did the debugging when it comes time to add the other sections!  =^D  

OK.  I can remove.

**Review by tri-adam — APPROVED** [review]

**Pull request merged**